### ISSUE TITLE
Ability to Return (Something, Unit) as output

### ISSUE THREAD

**Issue by joules-o** [softwaremill/tapir#365]
There's not a clear way to return `(Something, Unit)` as an endpoint output.

Constructing as 
```
Endpoint[..., ..., (H, Unit), Nothing]  = ...
.out(headerFn)
.out(emptyOutput).explicitly(tapir.typelevel.ParamConcat.concatSingleAndSingle)
```

 returns 

```
java.lang.IllegalStateException: Outputs and output values don't match in output:
Multiple(Vector(Mapped(Multiple(Vector(Header(...))),
sttp.tapir.EndpointOutput$$Lambda$1525/603127357@4ed041bb,
sttp.tapir.EndpointOutput$$Lambda$1586/12522552018@19da529e))), 
values: List(Header(a3770cc3-b45c-4f90-a663-ebd4f333a343), ())
```

Trying to manually map like

```
Endpoint[..., ..., (H, Unit), Nothing]  = ...
      .out(statusCode(StatusCode.NoContent))
      .mapOut(header => (header, ()))(tuple => tuple._1)
```
throws `java.lang.ClassCastException: H cannot be cast to scala.Tuple2`.

This would make it easier to make composable logic for 204 endpoints in particular so my Unit response from the central logic could be used as `O` in the reusable functions that add my headers, cookies, etc.

Would this be solvable with something like `.out(noBody)`, or possibly have the option for a 204 No Content to be an output that tuples Unit?

**Comment by adamw** [comment]
Having a single output mapped to a tuple will be problematic - as then the mapper assumes these are values for two outputs (which are represented as tuples).

So if you have only a single output/input, you'll need to wrap these in a custom type, if you want to keep the "tuple-nature". If you have more outputs/inputs, and things still don't work, please provide a full example :)

**Comment by joules-o** [comment]
Hm. Would it be more feasible to have the ability to return `(H, Nothing)` as the value? Wrapping the outout up in a single type doesn't allow me to enforce an empty body on any endpoint extending some `return204` base endpoint.

**Comment by adamw** [comment]
The problem is in the tuple, not the value. `Endpoint[(X, Y), ...]` corresponds to an endpoint with two inputs.

Maybe you can have a custom type which is parametrised? E.g. `Result[H, T]` where `T` could be `()` for the 204 case? Or otherwise please provide a fuller example of what you are trying to achieve with the base endpoint, maybe there's another approach as well.

Btw. `Nothing` wouldn't work as there are no instances of `Nothing`, hence also no instances of `(H, Nothing)`, etc :)

**Comment by adamw** [comment]
This should be fixed in 0.14.0. You can now map to tuples & unit types, as well as write generic funcitons which extend any endpoint with an input/output.
